src/ripple/rpc/impl/Role.cpp

```diff
@@ -104,36 +105,43 @@ requestInboundEndpoint (Resource::Manager& manager,
         forwardedFor);
 }
 
-std::string
+boost::string_view
 forwardedFor(http_request_type const& request)
 {
-    static std::string const forStr{"for="};
-    static std::size_t const forStrLen{forStr.size()};
-
     auto it = request.find("X-Forwarded-For");
     if (it != request.end())
     {
-        std::string forwardedFor = it->value().to_string();
-        std::size_t const pos = forwardedFor.find(',');
-        if (pos != forwardedFor.npos)
-            forwardedFor.resize(pos);
-        return forwardedFor;
+        return boost::beast::http::ext_list{
+            it->value()}.begin()->first;
     }
+
     it = request.find("Forwarded");
     if (it != request.end())
     {
-        std::stringstream ss(it->value().to_string());
-        std::string directive;
-        while (std::getline(ss, directive, ';'))
+        static std::string const forStr{"for="};
+        auto found = std::search(it->value().begin(), it->value().end(),
+            forStr.begin(), forStr.end(),
+            [](char c1, char c2)
+            {
+                return boost::beast::detail::ascii_tolower(c1) ==
+                       boost::beast::detail::ascii_tolower(c2);
+            }
+        );
+
+        if (found == it->value().end())
+            return boost::string_view{};
+
+        char const& val = *(found + forStr.size());
+        std::size_t n;
+        for (n = 0; n < std::distance(&val, it->value().end()); ++n)
         {
-            std::size_t const found = directive.find(forStr);
-            if (found == directive.npos)
-                continue;
-            return directive.substr(found + forStrLen);
+            if (*(&val + n) == ';')
+                break;
         }
+        return boost::string_view{&val, n};
```

Comment:
Consider this alternative:
```
        found += forStr.size();
        auto pos{it->value().find(';', forStr.size())};
        if (pos != boost::string_view::npos)
            return {found, pos + 1};
        return {found, static_cast<size_t>(std::distance(found, it->value().end()))};
```

Reply:
Thanks, @miguelportilla -- this is preferable.